### DISCUSSION TITLE
syncing sql server back up files

### DISCUSSION THREAD

**Discussion by goake** [abraunegg/onedrive#2314] [Q&A] [answered]
I'm really a novice when it comes to linux - but I'm sure this is a simple answer, just beyond me.

I have azure-sql-edge running in a docker container, and I want to copy my database back up files to one drive.

I've configured onedrive to successfully synch a single directory, 'DB_Backups'. and I have mapped this directory as a volume within the docker container. The first problem was that the azure-sql-edge didnt have permissions to write to this directory, so I changed the permisions to of 'DB_Backups' to 777 (I'm sure this isnt a great idea, but I'm fault finding). Now azure-sql-edge can succesfully write to this directory.

But onedrive complains with this error:
```
ERROR: The local file system returned an error with the following message:
  Error Message:    DB_Backups/azuresqledge/Renewables/FULL/azuresqledge_Renewables_FULL_20230224_144449.bak: Permission denied
```
I can fix this by changing the permissions on the file to 777, but there must be another way rather than doing that every time.

Thanks

**Comment by abraunegg** [comment, marked as answer]
@goake 
> I can fix this by changing the permissions on the file to 777, but there must be another way rather than doing that every time.

So the '777' is all about USER GROUP WORLD permissions. You want to most likely use 660 rather than 777. Use https://chmod-calculator.com/ to assist your understanding.

To resolve this, this is more of a Linux security question that something with the 'onedrive' client, however the following should help you get there.

From the 'azure-sql-edge' container, note what the USER and GROUP name is that is able to write the database backup file. Use `ls -la` to view this.

Using those details, have a look at `/etc/group` and match the GROUP name to a number. Now that you have this number, start tracing back in your 'onedrive' container what your 'DB_Backups' folder looks like.

Record what the USER and GROUP is, then go look at the `/etc/group` file again and see if that group is there - it potentially will not be.

You potentially then need to add a 'new group' to your 'onedrive' container:
```
groupadd -g GROUP mygroup
```
For example, if your GROUP name on the 'azure-sql-edge' is 'mssql' with a GID of 12345, and on the 'onedrive' container you can see that the folders / files are still owned by a group of '12345':
```
groupadd -g 12345 mysql
```
You then want to add the 'onedrive' user (or whatever the userid is that is running the user process under Docker) to this new 'mysql' group on your 'onedrive' container:
```
usermod -a -G mysql onedrive
```

I have not tested any of this - so your mileage may vary - but this is how you need to work through this permissions problem.

**Comment by goake** [comment]
Thanks for the answer, and I do appreciate that this is not a 'onedrive' client issue, but I'm going to persist asking here as I cant seem to find any answers with google!

Changing the permissions on the 'DB_Backups' to 660 results in this error when creating a sql backup:
`Cannot open backup device '/g/azuresqledge/Renewables/FULL/azuresqledge_Renewables_FULL_20230224_193518.bak'. Operating system error 5(Access is denied.).`

So I've put it back to 777 (for now)

I dont know if this makes any difference to your advice but my onedrive client is installed on the host (I couldnt get it working in a container, as soon as I authorised it immediately started synching everythging in my OneDrive, and I only want the one folder synched rather than the TBs of other stuff)

> From the 'azure-sql-edge' container, note what the USER and GROUP name is that is able to write the database backup file. Use ls -la to view this.

mssql    root

I dont understand this. The logfile for when the 'azure-sql-edge' container starts:
```
`Azure SQL Edge will run as non-root by default.
This container is running as user mssql.
Your master database file is owned by mssql.
```
But adding the user that onedrive client is running under to the root group does fix stuff. But surely this is not a smart thing to do??
`

**Reply by goake** [reply]
I did solve this. I know that this is not an issue with the OneDrive client, but since copying sql server backup files into OneDrive seems a usefull thing to do I've put my soloution here, bearing in mind I dont really know what I'm talking about...

I believe the issue is with the docker image from microsoft for azure-sql-edge. This runs as user mssql, but mssql does not belong to any group in this image. When this mssql user writes to the volume mounted on the host system, the group ownership of null becomes 0, ie root, so the user that onedrive is running under does not have permissions to this file.

I created a new image with this dockerfile:

```
FROM mcr.microsoft.com/azure-sql-edge AS release

        ENV ACCEPT_EULA=Y
        ENV MSSQL_SA_PASSWORD=SomeStrongPassword
        ENV MSSQL_PID=Premium

        USER root

        RUN groupadd -g 10001 mssql \
                && usermod -g mssql mssql \
                && chown -R mssql /var/opt/mssql

        USER mssql
```
Which basically creates a group for mssql, adds mssql to that group, and takes ownership of the directory needed to run.

Then on the host system, create a user and group for mssql:

```
sudo groupadd -g 10001 mssql
sudo useradd mssql -u 10001 -g 10001
```
Then added the onedrive user to the mssql group

`sudo usermod -a -G mssql onedriveuser`

Lastly, changed the ownership of the onedrive sync director to mssql - and then everything seems to work.

**Comment by abraunegg** [comment]
@goake 
From what you have done - this certainly seems the right way to solve your issue.
